### PR TITLE
Add support for Oauth2 Authentication using web browser

### LINKED ISSUE #40: oidc support for dbt-trino
A great and unique feature of Trino is that it allows users to authenticate through OIDC. This is not something commonly seen with databases (only Snowflake and Denodo support this feature). it's a much safer use as it doesn't require storing any local credentials.

I saw that this feature is implemented in the trino-python-client but not yet supported in the dbt adapter.

Some gotchas:
- As you can specify # of threads and each model usually generates lots of queries, it might be annoying to see lots of browser windows opening. So ideally the access token is cached and reused for every query.
- Some dbt jobs may take a long time, we need to ensure that queries don't fail because of token expiration while running a dbt project.

### PR DESCRIPTION
**Pull request by mdesmet**
Fixes #40 

The python client seems to print a url to follow for each query execution. I think it will be more user friendly to execute the default browser instead (similar to how snowflake does it).

image

### CODE REVIEW

File: dbt/adapters/trino/connections.py [hunk 1]
```diff
@@ -21,32 +22,201 @@
...
+        klazz = cls._create_trino_profile(data)
+        return klazz.validate(data)
+
+    @classmethod
+    def from_dict(cls, data: Any):
+        klazz = cls._create_trino_profile(data)
+        return klazz.from_dict(data)
+
+
+class HttpScheme(Enum):
```
Comment (hovaesco) [inline dbt/adapters/trino/connections.py:60]: nit: move above `TrinoCredentialsFactory`

File: test/unit/test_adapter.py [hunk 2]
```diff
@@ -7,6 +7,14 @@
 from dbt.clients import agate_helper
 
 from dbt.adapters.trino import TrinoAdapter
+from dbt.adapters.trino.connections import (
```
Comment (hovaesco) [inline test/unit/test_adapter.py:10]: `TrinoKerberosCredentials` missing in tests

File: dbt/adapters/trino/connections.py [hunk 3]
```diff
@@ -21,32 +22,201 @@
...
     host: str
     port: Port
     user: str
-    password: Optional[str] = None
-    jwt_token: Optional[str] = None
-    client_certificate: Optional[str] = None
-    client_private_key: Optional[str] = None
+    cert: Optional[str] = None
     method: Optional[str] = None
+    http_scheme: HttpScheme = HttpScheme.HTTP
```
Comment (hovaesco) [inline dbt/adapters/trino/connections.py:72]:
```python
    @property
    def http_scheme(self):
        return HttpScheme.HTTP

File: dbt/adapters/trino/connections.py [hunk 4]
```diff
@@ -21,32 +22,201 @@
...
+    user: str
+    password: str
+    cert: Optional[str] = None
+    http_headers: Optional[Dict[str, str]] = None
+    session_properties: Optional[Dict[str, Any]] = None
+    _ALIASES = {"catalog": "database"}
+
+    @property
+    def type(self):
+        return "trino"
```
Comment (hovaesco) [inline dbt/adapters/trino/connections.py:130]: This and `unique_field` is repeated in each Credentials class, is it feasible to have `TrinoCredentials` class and inherit from it in specific Credentials methods?

**Review by hovaesco — COMMENTED** [review]
Some comments, overall looks good.

File: dbt/adapters/trino/connections.py (hunk 3 above)
Comment (mdesmet) [reply dbt/adapters/trino/connections.py:72]: You could theoretically configure a Trino server with https but without any authentication configured. This is also documented as such.

File: dbt/adapters/trino/connections.py (hunk 4 above)
Comment (mdesmet) [reply dbt/adapters/trino/connections.py:130]: Will try, remember hitting some limitations with inheritance and `@dataclass` classes in the past.

Comment (mdesmet) [reply dbt/adapters/trino/connections.py]: Ok, that works, I think the only limitation is about the order of fields with a default value vs without default value but doesn't apply here, as long we don't try to put the properties in the base class.

File: README.md [hunk 5]
```diff
@@ -104,6 +104,7 @@ Please only use lower case schema names with this adapter.
 - [kerberos](https://trino.io/docs/current/security/kerberos.html) - Specify username in `user`
 - [jwt](https://trino.io/docs/current/security/jwt.html) - Specify JWT token in `jwt_token`
 - [certificate](https://trino.io/docs/current/security/certificate.html) - Specify a client certificate in `client_certificate` and private key in `client_private_key`
+- [oauth](https://trino.io/docs/current/security/oauth2.html) - Optionally install keyring to cache the OAuth2 token over multiple dbt invocations (install keyring using `pip install 'trino[external-authentication-token-cache]'`).
```
Comment (hovaesco) [inline README.md:107]: I would say that is recommended to install keyring, not optional. We could just mention that is optional.

**Review by hovaesco — APPROVED** [review]
LGTM, please improve a commit message `Support oauth connections using webbrowser popups` to `Support OAuth authentication using web browser`

**Review by hovaesco — APPROVED** [review]

**Pull request merged**